python/ray/ml/utils/remote_storage.py

```diff
@@ -0,0 +1,165 @@
+
+from ray import logger
+
+# We keep these constants for legacy compatibility with Tune's sync client
+# After Tune fully moved to using pyarrow.fs we can remove these.
+S3_PREFIX = "s3://"
+GS_PREFIX = "gs://"
+HDFS_PREFIX = "hdfs://"
+ALLOWED_REMOTE_PREFIXES = (S3_PREFIX, GS_PREFIX, HDFS_PREFIX)
+
+
+def _assert_pyarrow_installed():
+    if pyarrow is None:
+        raise RuntimeError(
+            "Uploading, downloading, and deleting from cloud storage "
+            "requires pyarrow to be installed. Install with: "
+            "`pip install pyarrow`. Subsequent calls to cloud operations "
+            "will be ignored."
+        )
+
+
+def fs_hint(uri: str) -> str:
+    """Return a hint how to install required filesystem package"""
+    if pyarrow is None:
+        return "Please make sure PyArrow is installed: `pip install pyarrow`."
+    if fsspec is None:
+        return "Try installing fsspec: `pip install fsspec`."
+
+    from fsspec.registry import known_implementations
+
+    protocol = urllib.parse.urlparse(uri).scheme
+
+    if protocol in known_implementations:
+        return known_implementations[protocol]["err"]
+
+    return "Make sure to install and register your fsspec-compatible filesystem."
+
+
+def is_non_local_path_uri(uri: str) -> bool:
+    """Check if target URI points to a non-local location"""
+    parsed = urllib.parse.urlparse(uri)
+    if parsed.scheme == "file" or not parsed.scheme:
+        return False
+
+    if bool(get_fs_and_path(uri)[0]):
+        return True
+    # Keep manual check for prefixes for backwards compatibility with the
+    # TrialCheckpoint class. Remove once fully deprecated.
+    if any(uri.startswith(p) for p in ALLOWED_REMOTE_PREFIXES):
+        return True
+    return False
+
+
+# Cache fs objects
+_cached_fs = {}
+
+
+def get_fs_and_path(
+    uri: str,
+) -> Tuple[Optional["pyarrow.fs.FileSystem"], Optional[str]]:
+    if not pyarrow:
+        return None, None
+
+    parsed = urllib.parse.urlparse(uri)
+    path = parsed.path
+
+    cache_key = (parsed.scheme, parsed.netloc)
+
+    if cache_key in _cached_fs:
+        fs = _cached_fs[cache_key]
+        return fs, path
+
+    try:
+        fs, path = pyarrow.fs.FileSystem.from_uri(uri)
+        _cached_fs[cache_key] = fs
+        return fs, path
+    except pyarrow.lib.ArrowInvalid:
+        # Raised when URI not recognized
+        if not fsspec:
+            # Only return if fsspec is not installed
+            return None, None
+
+    # Else, try to resolve protocol via fsspec
+    try:
+        fsspec_fs = fsspec.filesystem(parsed.scheme)
+    except ValueError:
+        # Raised when protocol not known
+        return None, None
+
+    fs = pyarrow.fs.PyFileSystem(pyarrow.fs.FSSpecHandler(fsspec_fs))
+    _cached_fs[cache_key] = fs
+
+    return fs, path
+
+
+def delete_at_uri(uri: str):
+    _assert_pyarrow_installed()
+
+    fs, bucket_path = get_fs_and_path(uri)
+    if not fs:
+        raise ValueError(
```

Comment:
I would like to keep it as is because we can raise more actionable error messages that way - i.e. include the operation ("Could not clear URI contents", "Could not upload to URI...") etc. Let me know if you feel strongly about this